### DISCUSSION TITLE
UnboundTypeError:  Prs3d.GetDeflection_1

### DISCUSSION THREAD

**Discussion by qouv** [donalffons/opencascade.js#138] [Q&A] [answered]
Hello,

When I try to use GetDeflection I get the error below. Do you have any idea how to solve this?

**UnboundTypeError: Cannot call Prs3d.GetDeflection_1 due to unbound types: 16NCollection_Vec3IdE**

**Comment by donalffons** [comment, marked as answer]
I believe that should work in the latest version `2.0.0-beta.ae74abe`. Instead of `Graphic3d_Vec3d` (as in the function signature), you should be able to use `Select3D_Vec3` instead (they're a differently named typedef's to the same underlying type).

**Reply by qouv** [reply]
In the XDE example of C++ code:
https://draw.sview.ru/xde-export-step2gltf/
The line is:
`anAlgo.ChangeParameters().Deflection = Prs3d::GetDeflection (aCompound, aDrawer);`

I couldnt find the correct parameters to use with  Prs3d.GetDeflection_1(); 

aCompound and aDrawer are still giving unboundtypeerror with the latest version` 2.0.0-beta.ae74abe`

**Reply by donalffons** [reply]
You're right. There's an issue with that binding. Will fix that soon in one of the next releases. For some context: OpenCascade defines the type `NCollection_Vec3<Standard_Real>` under several names: `Graphic3d_Vec3d`, `Select3D_Vec3`, `SelectMgr_Vec3`. When building with Emscripten, we can only bind that underlying type once (otherwise, there will be an error). Currently, all of those types are filtered out.

The code on that site you mentioned looks outdated - at least it doesn't match the current [OCCT docs](https://dev.opencascade.org/doc/refman/html/class_prs3d.html).

**Reply by donalffons** [reply]
This now works in `2.0.0-beta.44294b3`.
```js
const d = oc.Prs3d.GetDeflection_1(new oc.Graphic3d_Vec3d_3(0, 0, 0), new oc.Graphic3d_Vec3d_3(1, 1, 1), 0.1);
console.log(d); // 0.4
```

**Reply by qouv** [reply]
Wow so fast :D

I implemented it like this and works without exception:
```js
const BoundBox = new oc.Bnd_Box_1();
oc.BRepBndLib.Add(aCompound, BoundBox, false);
   
anAlgo.ChangeParameters().Deflection = oc.Prs3d.GetDeflection_2(BoundBox, aDrawer.MaximalChordialDeviation(), aDrawer.DeviationCoefficient());
```
I was thinking my conversion speed was about this. But nothing changed, it lasts forever for big parts.

How did this sview.ru guys achieved this speed? Do you have any idea, are they using high CPU server or etc. Because their speed is very okay for me.

At first I thought there must be an api or npm package for step2glb. But no :(

OCJS looks like the best option since I am developing nodejs app. I guess I cant run c++ code with my current knowledge.

**Reply by donalffons** [reply]
I believe this code might have no effect. Can you check if those changes get applied correctly?
```js
anAlgo.ChangeParameters().Deflection = 0.123;
console.log(anAlgo.ChangeParameters().Deflection); // might print something else than 0.123;
```
This problem is tracked in #125 and I'm still in the process of figuring out how a good workaround would look like.

For meshing, I would recommend to use this syntax instead until that problem is fixed:
```js
new oc.BRepMesh_IncrementalMesh_2(aCompound, linDeflection, false, angDeflection, true);
```

Generally, when there are things that currently cannot be implemented in JS, you should be able to work-around most problems with writing helper functions in C++ and creating a custom build.

Regarding your performance problems:

1. Did you do some actual time measurements? How much time does `initOpenCascade` / reading & transferring STEP / meshing / exporting to GLB take? If you're using NodeJS, maybe your problem is the time required for instantiation, i.e. `initOpenCascade`?!
2. Make sure you're using a recent version of NodeJS - the newer the better. Older NodeJS versions tend to suffer from extremely long startup times for large WASM projects. But even in newer NodeJS versions, startup times can be quite long. A good solution would be if NodeJS would be able to cache the compiled WASM, see [this issue](https://github.com/nodejs/node/issues/36671). But that still seems to be work-in-progress.
3. If you want you can share a STEP file here and I might have a look.

Generally if you want to convert large STEP files and you don't need to do that in a browser, C++ might be the better Option. OCCT has [an example](https://github.dev/Open-Cascade-SAS/OCCT/blob/b2a63f9b6fad9c9a5b23d15b1ad6778d02aec91c/samples/ios/UIKitSample/UIKitSample/OcctViewer.mm#L221) for that.

**Reply by qouv** [reply]
You are right.

`console.log(anAlgo.ChangeParameters().Deflection); ` prints 0.001 instead.

And this one line saved me:

`new oc.BRepMesh_IncrementalMesh_2(aCompound, linDeflection, false, angDeflection, true);`

I guess it will do until we hire someone writes C++. Now I will just work on optimizing GLB file for 3D web view. Thank you very much.

1. Triangulating was the one takes time. Initiation (initOpenCascade) takes around 30 sec. I will try to initiate at first with node app start.
2. I will update Node version, if you say so! And share here if I see any difference.
3. Not a specific file. I tried many and decided it was about size. Small parts were no problem.

I know there will be always performance differences between C++ and JS. I need a quick solution for this project and found it.  Also upgrade my AWS server to compute optimized one. 

You guys are doing great job implementing OCC to JS for the community. Keep going!

**Reply by donalffons** [reply]
Maybe as a point of reference: I'm working on a browser-based STEP file viewer using OpenCascade.js right now. Loading a 520MB file takes around 13 min on my fairly fast laptop (10 minutes of `StepCAFControl::Read` and `StepCAFControl::Translate` and 3 minutes of meshing). I believe that file size is pretty close to the maximum feasible size...

Support for native WASM exceptions landed in Emscripten a couple of days ago. Once that improvement is integrated into OpenCascade.js, I expect loading times to improve significantly.

Not to try to make this into a sales pitch, but if your company ever needs support in OpenCascade-related topics, you can [hire me](https://www.interactiveimpressions.com/contact) ;-).

**Reply by qouv** [reply]
I think time is not a problem anymore.  I am dealing with mesh count. Big files come up with over 10k+ meshes and creates unnecessary draw calls. I need to reduce mesh number somehow.

We are at a very early stage right now. But you are on my mind though. Maybe after the investment round :)
